python/cudf/cudf/core/dataframe.py

```diff
@@ -2638,31 +2678,134 @@ def set_index(self, index, drop=True):
-                df._drop_column(index)
-            return df.set_index(self[index])
-        # Otherwise
+            Whether to drop corresponding column for str index argument
+        append : boolean
+            Append current to new index to form a multiindex
+        inplace : boolean
+            Modify the DataFrame in place (do not create a new object)
+        verify_integrity : boolean
+            Check for duplicates in the new index
+
+        Examples
+        --------
+        Set existing column as index
+        >>> df = cudf.DataFrame({"a": [1,2,3,4,5],
+        ... "b":["a", "b", "c", "d","e"],
+        ... "c":[1.0, 2.0, 3.0, 4.0, 5.0]})
+        >>> df.set_index(["a", "b"])
+               c
+        a b
+        1 a  1.0
+        2 b  2.0
+        3 c  3.0
+        4 d  4.0
+        5 e  5.0
+
+        Set new Index instance as index
+        >>> df.set_index(cudf.RangeIndex(10, 15))
+            a  b    c
+        10  1  a  1.0
+        11  2  b  2.0
+        12  3  c  3.0
+        13  4  d  4.0
+        14  5  e  5.0
+
+        Setting `append=True` will combine current index with the column
+        >>> df.set_index("a", append=True)
+             b    c
+          a
+        0 1  a  1.0
+        1 2  b  2.0
+        2 3  c  3.0
+        3 4  d  4.0
+        4 5  e  5.0
+
+        `inplace` operation:
+        >>> df.set_index("a", inplace=True)
+        >>> df
+           b    c
+        a
+        1  a  1.0
+        2  b  2.0
+        3  c  3.0
+        4  d  4.0
+        5  e  5.0
+        """
+
+        if not isinstance(index, list):
+            index = [index]
+
+        # Preliminary type check
+        col_not_found = []
+        columns_to_add = []
+        names = []
+        to_drop = []
+        for i, col in enumerate(index):
+            # Is column label
+            if is_scalar(col) or isinstance(col, tuple):
+                if col in self.columns:
+                    columns_to_add.append(self[col])
+                    names.append(col)
+                    if drop:
+                        to_drop.append(col)
+                else:
+                    col_not_found.append(col)
+            else:
+                # Try coerce into column
+                if not is_column_like(col):
+                    try:
+                        col = as_column(col)
+                    except TypeError:
+                        raise f"{col} cannot be converted to column-like."
+                if isinstance(col, (cudf.MultiIndex, pd.MultiIndex)):
+                    col = (
+                        cudf.from_pandas(col)
+                        if isinstance(col, pd.MultiIndex)
+                        else col
+                    )
+                    cols = [col._data[x] for x in col._data]
+                    columns_to_add = columns_to_add + cols
+                    names = names + col.names
+                else:
+                    # Column-like
+                    columns_to_add.append(col)
```

Comment:
```suggestion
                    columns_to_add.append(as_column(col))
```

`col` can be a pandas object too, in which case we want it on GPU memory.

Reply:
This actually introduced a corner case where `RangeIndex` does not need to be instantiated in some situations. Updated a fix.